cdap-app-templates/cdap-etl/hydrator-spark-core-base/src/main/java/io/cdap/cdap/etl/spark/batch/RDDCollection.java

```diff
@@ -176,9 +179,28 @@ public RDDCollection(JavaSparkExecutionContext sec, FunctionCache.Factory functi
 
       Iterator<Column> leftIter = leftJoinColumns.iterator();
       Iterator<Column> rightIter = rightJoinColumns.iterator();
-      Column joinOn = eq(leftIter.next(), rightIter.next(), joinRequest.isNullSafe());
-      while (leftIter.hasNext()) {
-        joinOn = joinOn.and(eq(leftIter.next(), rightIter.next(), joinRequest.isNullSafe()));
+
```

Comment:
nit: Consider rewriting it with streams, it would look nicer than iterators, e.g.
```
if (seenRequired) {
 joinOn = intStream.range(0, leftJoinColumns.size())
 .map(i -> eq(leftJoinColumns.get(i), rightJoinColumns.get(i), joinRequest.isNullSafe())
 .reduce((a, b) -> a.and(b)).get()
} else {
 joinOn = intStream.range(0, leftJoinColumns.size())
 .map(i -> {
 collectLeftJoinOnCols(listOfListOfLeftCols, i, leftJoinColumns.get(i));
 return eq(getLeftJoinOnCoalescedColumn(leftJoinOnCurrent, index, listOfListOfLeftCols), rightJoinColumns.get(i), joinRequest.isNullSafe()));
})
 .reduce((a, b) -> a.and(b)).get();
}
```
Note: the code is example and untested

Reply:
Thanks! implemented the streams approach and tested.